### PR TITLE
Crafting GUI: Filter recipes by what can fit into the result (longest_side, volume, mass)

### LINKED ISSUE #74635: Filter items with pockets by pocket's size, len, etc.
### Is your feature request related to a problem? Please describe.

It is very tedious whether I have an item with a pocket that can contain a 78 cm long item (pickaxe). Or if I can craft such an item.

### Solution you would like.

Add a filter such as `l:X` (l for length) to filter items with a pocket that can contain items that are X cm long.

This isn't perfect by itself, but it would filter out all the clothing etc. on the tile that definitely cannot contain items with selected lengths.

### Describe alternatives you have considered.

Currently one of the ways is to read this menu in the item description and search for every item there, whether I have them in my base and then search again, if I can craft this item:
image

This method is feasible for plank:
image

Otherwise, if the menu text (Can be stored in) would highlight items that are nearby, I would only need to search the crafting menu, if I don't have any item nearby.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Brambor**
#### Summary
Interface "Crafting GUI: Filter recipes by what can fit into the result (longest_side, volume, mass)"

#### Purpose of change

 - Part of #74635 - crafting GUI, not AIM (That is for my next PR.)

#### Describe the solution

image

1. add filters `L:`, `V:`, `M:` to crafting GUI
1. Parse user input by an existing parser for units. It is the same parser that is used for loading `itype`s from JSON.
1. If an error is thrown on invalid input, the user is notified with a helpful message:
   - image
1. Create a fake item, set its longest_side/volume/mass to specified in the filter.
1. For each recipe try fitting the fake item into the recipe `item.can_contain(fake_item)`.
1. fix in `item.cpp`: `units::mass_max * 1.0` resulted in `-units::mass_max`. Fixed by avoiding multiplying by `1.0`.

#### Describe alternatives you've considered

_... I wrote a detailed thing there, where the hell is it?_

#### Testing

 - Use all the filters in crafting GUI, try all the suggested "Valid examples". Note: `2m` is not valid `2meter` is. Maybe not to be confused with `minutes`?
 - Try negative numbers. Like `M: -250g`.
   - Works for mass and longest_side - all recipes are shown.
   - Doesn't work for volume - none are shown. This is ok.

### CODE REVIEW

**Comment by db48x** [comment]
I like the idea. Can you add a comment documenting what the assert is checking for?

Also, why remove wildcards from an unrelated query type?

**Comment by Brambor** [comment]
> Can you add a comment documenting what the assert is checking for?

I went with a different approach and am finalizing it now. The assert will not be there.

> Also, why remove wildcards from an unrelated query type?

I added that query type a while back and didn't know better. Now I know and I want to make it in line with the other filters.

It can be a separate PR, I aggregated it since it is close by.

**Comment by db48x** [comment]
Probably better to have a separate pull request for that then.

**Comment by Brambor** [comment]
I updated to the new approach. I still need to do the inventory filter, which will have to be a bit different since I cannot throw that helpful error on every keystroke (the inventory filter is live).

**Comment by Brambor** [comment]
This does have a merge conflict with #78423. I am happy to rebase one of the PRs once one of them gets merged.

File: data/json/items/fake.json [hunk 1]
```diff
@@ -481,6 +481,14 @@
     "copy-from": "fake_appliance_tool",
     "charges_per_use": 20
   },
+  {
+    "id": "debug_item_search",
+    "name": { "str": "debug item search" },
```
Comment (Night-Pryanik) [inline data/json/items/fake.json:486]:
```suggestion
    "name": { "str": "debug item search", "//~": "NO_I18N" },
```

File: data/json/items/fake.json [hunk 2]
```diff
@@ -481,6 +481,14 @@
...
+    "type": "GENERIC",
+    "symbol": "?",
+    "description": "This is a dummy item used by item search.  You should never see it outside of debug.",
```
Comment (Night-Pryanik) [inline data/json/items/fake.json:489]:
```suggestion
   "description": { "str": "This is a dummy item used by item search.  You should never see it outside of debug.", "//~": "NO_I18N" },
```

File: src/recipe_dictionary.cpp [hunk 3]
```diff
@@ -328,6 +337,71 @@ std::vector<const recipe *> recipe_subset::search(
 
     // prepare search
     switch( key ) {
+        case search_type::length: {
+            auto const error = [txt]( char const *, size_t /* offset */ ) {
+                throw math::runtime_error( _( string_format(
+                                                  "Failed to convert '%s' to length.\nValid examples:\n122 cm\n1101mm\n2   meter",
```
Comment (Night-Pryanik) [inline src/recipe_dictionary.cpp:343]: These three spaces before `meter` are intended?

Comment (Brambor) [reply src/recipe_dictionary.cpp:343]: Yes. In these examples I am trying to show off that spacing does not matter.

Comment (Brambor) [reply src/recipe_dictionary.cpp:344]: I added a comment to the first occurrence of this.

**Comment by Brambor** [comment]
Huh, didn't know a special name was requested:
image

It should have said so in the last error, when it asked me to make it a global variable!

**Pull request merged**